test/python/transpiler/test_remove_final_measurements.py

```diff
@@ -0,0 +1,221 @@
+
+        # measure into all clbits of c0
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+
+        dag = circuit_to_dag(qc)
+        RemoveFinalMeasurements().run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+
+    def test_register_kept_if_measured_clbit_busy(self):
+        """
+        A register is kept if the measure destination bit is still
+        busy after measure removal.
+        """
+        c0 = ClassicalRegister(1)
+        qc = QuantumCircuit(QuantumRegister(1), c0)
+
+        # make c0 busy
+        qc.x(0).c_if(c0, 0)
+
+        # measure into c0
+        qc.measure(0, c0)
+
+        dag = circuit_to_dag(qc)
+        RemoveFinalMeasurements().run(dag)
+
+        self.assertSetEqual(set(dag.cregs.values()), {c0})
+        self.assertSetEqual(set(dag.clbits), set(c0))
+
+    def test_multi_bit_register_kept_if_not_measured_clbit_busy(self):
+        """
+        A multi-bit register is kept if it contains a busy bit even if
+        the measure destination bit itself is idle.
+        """
+        c0 = ClassicalRegister(2)
+        qc = QuantumCircuit(QuantumRegister(1), c0)
+
+        # make c0[0] busy
+        qc.x(0).c_if(c0[0], 0)
+
+        # measure into not busy c0[1]
+        qc.measure(0, c0[1])
+
+        dag = circuit_to_dag(qc)
+        RemoveFinalMeasurements().run(dag)
+
+        # c0 should not be removed because it has busy bit c0[0]
+        self.assertSetEqual(set(dag.cregs.values()), {c0})
+
+        # note: c0[1] should not be removed even though it is now idle
+        # because it is referenced by creg c0.
+        self.assertSetEqual(set(dag.clbits), set(c0))
+
+    def test_multi_bit_register_removed_if_all_bits_idle(self):
+        """A multibit register is removed when all bits are idle."""
+        qc = QuantumCircuit(1, 2)
+
+        # measure into single bit c0[0] of c0
+        qc.measure(0, 0)
+
+        dag = circuit_to_dag(qc)
+        RemoveFinalMeasurements().run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+
+    def test_multi_reg_shared_bits_removed(self):
+        """All registers sharing removed bits should be removed."""
+        q0 = QuantumRegister(2)
+        c0 = ClassicalRegister(2)
+        qc = QuantumCircuit(q0, c0)
+
+        # Create reg with shared bits (same as c0)
+        c1 = ClassicalRegister(bits=qc.clbits)
+        qc.add_register(c1)
+
+        # measure into all clbits of c0
+        qc.measure(0, c0[0])
+        qc.measure(1, c0[1])
+
+        dag = circuit_to_dag(qc)
+        remove_final_meas = RemoveFinalMeasurements()
+        remove_final_meas.run(dag)
+
+        self.assertFalse(dag.cregs)
+        self.assertFalse(dag.clbits)
+
+    def test_final_measures_share_dest(self):
+        """Multiple final measurements use the same clbit."""
+
+        def expected_dag():
+            qc = QuantumCircuit(QuantumRegister(2, "q0"))
+            return circuit_to_dag(qc)
```

Comment:
For this particular test, there's not really much point, other than for consistency with the other tests where it could be nice. I thought to do this because the variables in these functions should not be available to the outer function, so we can use the same intermediate variable names without risk of clobbering those of the "actual" DAG. The reverse isn't true though, apparently. The `expected_dag` function can see values assigned even lexically below it in the outer function, as long as those assignments run before `expected_dag`.

Reply:
^ though if you _use_ variables lexically defined below a function, you get shot out of the cannon haha